### ISSUE TITLE
Minibatch gradients aren't readily accessible in zero optimization level 1

### ISSUE THREAD

**Issue by bcui19** [microsoft/DeepSpeed#2329]
**Describe the bug**
For a given minibatch, using stage 1 zero optimization, the gradients aren't readily accessible. Say there are 'n' gradient accumulation steps, up to 'n-1' steps by the engine the gradients are accessible, but at the 'nth' call to '.backward' the gradients in the computation graph are gone. 

**To Reproduce**
I've included a minimal code snippet to reproduce the behavior:

```
import torch 
import os
import deepspeed
import datetime

def get_xdist_worker_id():
    xdist_worker = os.environ.get('PYTEST_XDIST_WORKER', None)
    if xdist_worker is not None:
        xdist_worker_id = xdist_worker.replace('gw', '')
        return int(xdist_worker_id)
    return None


def get_master_port():
    master_port = os.environ.get('DS_TEST_PORT', '29503')
    xdist_worker_id = get_xdist_worker_id()
    if xdist_worker_id is not None:
        master_port = str(int(master_port) + xdist_worker_id)
    return master_port


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.emb = torch.nn.EmbeddingBag(10, 3, mode="sum", sparse=True)
        self.linear = torch.nn.Linear(3, 1)

    def forward(self, x, offsets):
        # lat = self.emb(x, offsets)
        # print ("embedding now is: ", lat, lat.shape)
        return self.linear(self.emb(x, offsets))

class Adam(torch.optim.Optimizer):
    def __init__(self, dense_params, sparse_params):
        super().__init__(dense_params + sparse_params, defaults={})
        self.adam = torch.optim.Adam(dense_params)
        self.adam_sparse = torch.optim.SparseAdam(sparse_params)

    @torch.no_grad()
    def step(self, closure=None):
        loss_1 = self.adam.step(closure)
        loss_2 = self.adam_sparse.step(closure)

        if loss_1 is not None and loss_2 is not None:
            return loss_1 + loss_2
        return loss_1 or loss_2

def get_model_optimizer():
    torch.manual_seed(0)
    model = Model()
    optimizer = Adam(list(model.linear.parameters()), list(model.emb.parameters()))
    return model, optimizer


def get_data(device):
    x = torch.tensor([1, 2, 4, 5, 4, 3, 2, 9], dtype=torch.long, device=device)
    offsets = torch.tensor([0, 4], dtype=torch.long, device=device)
    y = torch.tensor([[1.0], [0.0]], device=device)
    return x, offsets, y

def main():

    backend = "nccl"
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = get_master_port()
    os.environ['LOCAL_RANK'] = str(0)
    # NOTE: unit tests don't support multi-node so local_rank == global rank
    os.environ['RANK'] = str(0)
    os.environ['WORLD_SIZE'] = str(1)
    deepspeed.init_distributed(backend)

    deepspeed_config = {'zero_optimization': {'stage': 1}, "train_batch_size": 2, "steps_per_print": 1, "sparse_gradients": True, 'zero_allow_untested_optimizer': True, 'gradient_accumulation_steps': 2}

    model, optimizer = get_model_optimizer()
    loss = torch.nn.BCEWithLogitsLoss()
    engine, _, _, _ = deepspeed.initialize(model=model,
                                           optimizer=optimizer,
                                           config_params=deepspeed_config)

    x, offsets, y = get_data(engine.device)
    res = engine(x, offsets)

    engine.backward(loss(res, y))

    for k, v in engine.named_parameters():
        grad = v.grad
        print ("grad is: ", k, grad)

    engine.step()

    print ("after first step")
    print ()

    x, offsets, y = get_data(engine.device)
    res = engine(x, offsets)
    engine.backward(loss(res, y))

    for k, v in engine.named_parameters():
        grad = v.grad
        print ("grad is: ", k, grad)

if __name__ == "__main__":
    main()
```

**Expected behavior**
In this case, because 'gradient_accumulation_steps=2' I'd expect after the second backwards call I'd still have access to the gradients in the computation graph. 

**ds_report output**
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
sparse_attn ............ [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
transformer_inference .. [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/workdisk/brandon/composer/lib/python3.9/site-packages/torch']
torch version .................... 1.10.1+cu111
torch cuda version ............... 11.1
nvcc version ..................... 11.6
deepspeed install path ........... ['/workdisk/brandon/composer/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.5.10, unknown, unknown
deepspeed wheel compiled w. ...... torch 1.12, cuda 10.2
```

**System info (please complete the following information):**
 - OS: Ubunto 20.04.4
 - GPU count and types: 2 NVIDIA A-100s
 - Python version 3.9.14
 - Any other relevant info about your setup

**Launcher context**
In this case the above python script should be able to provide a minimal repo of the issue.

**Comment by loadams** [comment]
Thanks, I was able to use your code to repro this - taking a look at it.
